core/js/src/main/scala/com/avsystem/commons/serialization/nativejs/NativeJsonInput.scala

```diff
@@ -0,0 +1,160 @@
+package com.avsystem.commons
+package serialization.nativejs
+
+import com.avsystem.commons.annotation.explicitGenerics
+import com.avsystem.commons.serialization.GenCodec.ReadFailure
+import com.avsystem.commons.serialization.*
+import com.avsystem.commons.serialization.json.RawJson
+
+import scala.scalajs.js
+import scala.scalajs.js.JSON
+
+class NativeJsonInput(value: js.Any, options: NativeFormatOptions) extends InputAndSimpleInput { self =>
+  private def read[T](expected: String)(matcher: PartialFunction[Any, T]): T =
+    matcher.applyOrElse(value, (_: Any) => throw new ReadFailure(s"$expected expected."))
+
+  override def readNull(): Boolean =
+    value == null
+
+  override def readString(): String =
+    read("String") {
+      case s: String => s
+    }
+
+  override def readDouble(): Double =
+    read("Double") {
+      case v: Double => v
+    }
+
+  override def readInt(): Int =
+    read("Int") {
+      case v: Int => v
+    }
+
+  override def readLong(): Long = {
+    def longFromString(s: String): Long =
+      try s.toLong
+      catch {
+        case e: NumberFormatException => throw new ReadFailure(s"Cannot read Long", e)
+      }
+    (value: Any) match {
+      case s: String => longFromString(s)
+      case i: Int => i
+      case d: Double if d.isWhole => d.toLong
+      case b: js.BigInt => longFromString(b.toString)
+      case b if js.typeOf(b) == "bigint" => longFromString(b.asInstanceOf[js.BigInt].toString)
```

Comment:
Added, for some reason unknown to me, type pattern match does not work as expected

Reply:
I initially thought it might be a matter of the example, since runtime type tests are based on values, not types in Scala.js, but that does not seem to be the issue.